source/hic_hal/stm32/stm32h743ii/extended_features/source/udb_extended_features_task.c

```diff
@@ -0,0 +1,55 @@
+#include "udb_extended_features_task.h"
+#include "util.h"
+#include "adc.h"
+#include "adapter_detector.h"
+#include "rtx_os.h"
+
+static osThreadId_t s_udb_extended_features_task_id;
+static uint32_t s_udb_extended_features_task_cb[WORDS(sizeof(osRtxThread_t))];
+static uint64_t s_udb_extended_features_task_stack[UDB_EXTENDED_FEATURES_TASK_STACK / sizeof(uint64_t)];
+static const osThreadAttr_t k_udb_extended_features_task_attr =
+{
+    .name = "aux",
+    .cb_mem = s_udb_extended_features_task_cb,
+    .cb_size = sizeof(s_udb_extended_features_task_cb),
+    .stack_mem = s_udb_extended_features_task_stack,
+    .stack_size = sizeof(s_udb_extended_features_task_stack),
+    .priority = UDB_EXTENDED_FEATURES_TASK_PRIORITY,
+};
+
+static uint32_t s_timer_3s_cb[WORDS(sizeof(osRtxTimer_t))];
+static const osTimerAttr_t k_timer_3s_attr = {
+    .name = "3s",
+    .cb_mem = s_timer_3s_cb,
+    .cb_size = sizeof(s_timer_3s_cb),
+};
+
+void timer_task_3s(void * arg)
+{
+    osThreadFlagsSet(s_udb_extended_features_task_id, UDB_EXTENDED_FEATURES_TASK_FLAGS_3S);
+}
+
+static void udb_extended_features_task(void *arg)
+{
+    uint16_t flags = 0;
+
+    adc_init();
```

Comment:
Don't we read the HW ID before this point? Doesn't the HW ID require the ADC? We should probably read it after initializing the ADC.

Reply:
ADC is only used for adapter detector. HW ID is detected by GPIO read.

Reply:
Really? I'm not sure if I have the correct schematics, but it seems like proto2 has a voltage divider to express the version and is connected to an ADC input. I guess if we just want to distinguish proto1 and proto2, a GPIO could work (although I don't quite understand how it can work at all in proto1 regardless, since the pin is floating).

I suppose addressing the hardware version reading can be done later, but maybe it would be good to pre-emptively move the HW version read to after ADC init, since at some point, at the very minimum when the next hardware revision after proto2 comes out, it's going to need to use the ADC.

Reply:
ADC1_2_3_EXTI15_R is just for ADC external trigger and can not be used as ADC input. So what Eric came up with is that we can enable internal weak pull-up register for the GPIO to make proto1 read high and proto2 read low. But you are right that this can only differentiate p1 and p2 and can not be used for future versions.

Reply:
How is this going to work for future versions? Are they going to move to a different pin?

This sounds like it's going to be a headache for builds after p2. Anyway, I think this is a discussion for another time.